accord-core/src/main/java/accord/utils/SortedArrays.java

```diff
@@ -0,0 +1,535 @@
+                }
+            }
+        }
+        return found >= 0 ? found : -1 - to;
+    }
+
+    public static <T1, T2 extends Comparable<T1>> long findNextIntersectionWithOverlaps(T1[] as, int ai, T2[] bs, int bi)
+    {
+        return findNextIntersectionWithOverlaps(as, ai, bs, bi, (a, b) -> -b.compareTo(a), Comparable::compareTo);
+    }
+
+    public static <T1, T2> long findNextIntersectionWithOverlaps(T1[] as, int ai, T2[] bs, int bi, AsymmetricComparator<T1, T2> cmp1, AsymmetricComparator<T2, T1> cmp2)
+    {
+        return findNextIntersection(as, ai, bs, bi, cmp1, cmp2, Search.CEIL);
+    }
+
+    public static <T extends Comparable<? super T>> long findNextIntersection(T[] as, int ai, T[] bs, int bi)
+    {
+        return findNextIntersection(as, ai, bs, bi, Comparable::compareTo);
+    }
+
+    public static <T> long findNextIntersection(T[] as, int ai, T[] bs, int bi, AsymmetricComparator<T, T> comparator)
+    {
+        return findNextIntersection(as, ai, bs, bi, comparator, comparator, Search.FAST);
+    }
+
+    /**
+     * Given two sorted arrays, find the next index in each array containing an equal item.
+     *
+     * Works with CEIL or FAST; FAST to be used if precisely one match for each item in either list, CEIL if one item
+     *
+     * in either list may be matched to multiple in the other list.
+     */
+    private static <T1, T2> long findNextIntersection(T1[] as, int ai, T2[] bs, int bi, AsymmetricComparator<T1, T2> cmp1, AsymmetricComparator<T2, T1> cmp2, Search op)
+    {
+        if (ai == as.length)
+            return -1;
+
+        while (true)
+        {
+            bi = SortedArrays.exponentialSearch(bs, bi, bs.length, as[ai], cmp1, op);
+            if (bi >= 0)
+                break;
+
+            bi = -1 - bi;
+            if (bi == bs.length)
+                return -1;
+
+            ai = SortedArrays.exponentialSearch(as, ai, as.length, bs[bi], cmp2, op);
+            if (ai >= 0)
+                break;
+
+            ai = -1 - ai;
+            if (ai == as.length)
+                return -1;
+        }
+        return ((long)ai << 32) | bi;
+    }
+
+    public static <T extends Comparable<? super T>> int[] remapper(T[] src, T[] trg, boolean trgIsKnownSuperset)
+    {
+        return remapper(src, src.length, trg, trg.length, trgIsKnownSuperset);
+    }
+
+    /**
+     * Given two sorted arrays, where one is a subset of the other, return an int[] of the same size as
+     * the {@code src} parameter, with the index within {@code trg} of the corresponding element within {@code src},
+     * or -1 otherwise.
+     * That is, {@code src[i] == -1 || src[i].equals(trg[result[i]])}
+     */
+    public static <T extends Comparable<? super T>> int[] remapper(T[] src, int srcLength, T[] trg, int trgLength, boolean trgIsSuperset)
+    {
+        if (src == trg || (trgIsSuperset && trgLength == srcLength)) return null;
+        int[] result = new int[srcLength];
```

Comment:
needs a `Arrays.fill(result, -1);` else we put `0` as the case where trg ends *before* src, in my PR I do this and created property tests to show this.